app/shared/app-data/src/commonMain/kotlin/domain/torrent/TorrentServiceConnection.kt

```diff
@@ -0,0 +1,182 @@
+        scope.launch {
+            lifecycle.repeatOnLifecycle(Lifecycle.State.RESUMED) {
+                isAtForeground.value = true
+                try {
+                    if (!isServiceConnected.value) {
+                        logger.debug { "Lifecycle resume: Service is not connected, start connecting..." }
+                        startServiceWithRetry()
+                    }
+
+                    awaitCancellation()
+                } catch (_: CancellationException) {
+                    isAtForeground.value = false
+                    logger.debug { "Lifecycle pause: App moved to background." }
+                }
+            }
+        }
+    }
+    
+    /**
+     * 服务已断开连接, 通信对象变为不可用.
+     * 如果目前 APP 还在前台, 就要尝试重启并重连服务.
+     */
+    fun onServiceDisconnected() {
+        scope.launch(CoroutineName("TorrentServiceConnection - Lifecycle")) {
+            if (!isServiceConnected.value) {
+                // 已经是断开状态，直接忽略
+                return@launch
+            }
+            isServiceConnected.value = false
+            binderDeferred.cancel(CancellationException("Service disconnected."))
+            binderDeferred = CompletableDeferred()
+
+            // 若应用仍想要连接，则重新启动
+            if (isAtForeground.value) {
+                logger.debug { "Service is disconnected and app is in foreground, restarting service connection in 3s..." }
+                startServiceWithRetry()
+            } else {
+                logger.debug { "Service is disconnected and app is in background." }
+            }
+        }
+    }
+
+    private suspend fun startServiceWithRetry(
+        maxAttempts: Int = Int.MAX_VALUE, // 可根据需求设置
+        delayMillisBetweenAttempts: Long = 2500
+    ) {
+        var attempt = 0
+        while (attempt < maxAttempts && isAtForeground.value && !isServiceConnected.value) {
+            val binder = startServiceLock.withLock {
+                if (!isAtForeground.value || isServiceConnected.value) {
+                    logger.debug { "Service is already connected or app is not at foreground." }
+                    return
+                }
+                startService()
+            }
+            if (binder == null) {
+                logger.warn { "[#$attempt] startService() returned null binder, retry after $delayMillisBetweenAttempts ms" }
+                attempt++
+                delay(delayMillisBetweenAttempts)
+            } else {
+                logger.debug { "Service connected successfully: $binder" }
+                if (binderDeferred.isCompleted) {
+                    binderDeferred = CompletableDeferred()
+                }
+                binderDeferred.complete(binder)
+                isServiceConnected.value = true
+                return
+            }
+        }
+        if (!isServiceConnected.value) {
+            logger.error { "Failed to connect service after $maxAttempts retries." }
+        }
+    }
+
+    fun close() {
+        scope.launch {
```

Comment:
```suggestion
        scope.launch(NonCancellable) {
```

"清理" 工作比较适合 NonCancellable. 即使 scope 已被 cancel 也会执行.